source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -48,66 +50,121 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
+      if (common_response.has_header_mutation()) {
+        MutationUtils::applyHeaderMutations(common_response.header_mutation(), *request_headers_);
+      }
+    }
+    request_state_ = FilterState::IDLE;
+    decoder_callbacks_->continueDecoding();
+    return true;
+  }
+  return false;
+}
+
+bool Filter::handleResponseHeadersResponse(const HeadersResponse& response) {
+  if (response_state_ == FilterState::HEADERS) {
+    ENVOY_LOG(debug, "applying response_headers response");
+    if (response.has_response()) {
+      const auto& common_response = response.response();
+      if (common_response.has_header_mutation()) {
+        MutationUtils::applyHeaderMutations(common_response.header_mutation(), *response_headers_);
+      }
+    }
+    response_state_ = FilterState::IDLE;
+    encoder_callbacks_->continueEncoding();
+    return true;
+  }
+  return false;
+}
+
+bool Filter::handleImmediateResponse(const ImmediateResponse& response) {
+  if (response_state_ == FilterState::HEADERS) {
+    // Waiting for a response headers response, so return immediately now.
+    // Do this first in case both are in progress.
+    // We don't want to process any more stream messages after this.
+    // Close the stream before sending because "sendLocalResponse" triggers
+    // additional calls to this filter.
+    response_state_ = FilterState::IDLE;
+    closeStream();
+
+    ENVOY_LOG(debug, "Returning immediate response from processor on encoding path");
+    sendImmediateResponse(response, false);
+
+    return true;
+
+  } else if (request_state_ == FilterState::HEADERS) {
+    ENVOY_LOG(debug, "Returning immediate response from processor on decoding path");
+    request_state_ = FilterState::IDLE;
+    closeStream();
+    sendImmediateResponse(response, true);
+    return true;
+  }
+
+  return false;
+}
+
 void Filter::onGrpcError(Grpc::Status::GrpcStatus status) {
   ENVOY_LOG(debug, "Received gRPC error on stream: {}", status);
-  stream_closed_ = true;
   stats_.streams_failed_.inc();
+
   if (config_->failureModeAllow()) {
     // Ignore this and treat as a successful close
     onGrpcClose();
     stats_.failure_mode_allowed_.inc();
+
   } else {
-    // Use a switch here now because there will be more than two
-    // cases very soon.
-    switch (request_state_) {
-    case FilterState::HEADERS:
-      request_state_ = FilterState::IDLE;
-      decoder_callbacks_->sendLocalReply(
-          Http::Code::InternalServerError, "", nullptr, absl::nullopt,
-          absl::StrFormat("%s: gRPC error %i", kErrorPrefix, status));
-      break;
-    default:
-      // Nothing else to do
-      break;
-    }
+    stream_closed_ = true;
+    ImmediateResponse error_response;
+    error_response.mutable_status()->set_code(envoy::type::v3::StatusCode::InternalServerError);
+    error_response.set_details(absl::StrFormat("%s: gRPC error %i", kErrorPrefix, status));
+    handleImmediateResponse(error_response);
```

Comment:
Is the only situation in which you can send this synthetic error response when we're in the HEADERS request/response state? I'm thinking there might be a broader set of states...

Reply:
In the current codebase, "handleImmediateResponse" only needs to work in the "HEADERS" state. When we implement body processing two PRs from now, that will expand. However, there's a lot I need to experiment with regarding how Envoy handles "sendLocalReply" on both code paths before getting it right.

Reply:
Why do we need to wait until we are in response HEADERS to send an error message? I think as long as we haven't sent any response headers, we can do a `sendLocalReply`. We don't need to wait for `encodeHeaders` to do this. My original comment is that we should be able to do a `sendLocalReply` in any situation before we have sent the response headers back.

If you look at the behavior of something like `ConnectionManagerImpl::ActiveStream::onIdleTimeout()`, this has conditional behavior on whether we've processed response headers or not.